### PR TITLE
Update Google job store

### LINKED ISSUE #3388: Update Google job store
It looks like there are some errors when trying to run a workflow using the Google job store:

```
Traceback (most recent call last):
  File "sort.py", line 259, in <module>
    main()
  File "sort.py", line 247, in main
    sortedFileID = workflow.start(Job.wrapJobFn(setup,
  File "/Users/jonathanxu/toil/src/toil/common.py", line 845, in start
    rootJobDescription = rootJob.saveAsRootJob(self._jobStore)
  File "/Users/jonathanxu/toil/src/toil/job.py", line 2186, in saveAsRootJob
    self._saveJobGraph(jobStore, saveSelf=True)
  File "/Users/jonathanxu/toil/src/toil/job.py", line 2169, in _saveJobGraph
    jobStore.create(job.description)
  File "/Users/jonathanxu/toil/src/toil/jobStores/googleJobStore.py", line 182, in create
    self._writeString(jobDescription.jobStoreID, pickle.dumps(jobDescription, protocol=pickle.HIGHEST_PROTOCOL))
  File "/Users/jonathanxu/toil/src/toil/jobStores/googleJobStore.py", line 456, in _writeString
    self._writeFile(jobStoreID, StringIO(stringToUpload), **kwarg)
TypeError: initial_value must be str or None, not bytes
```
**Steps to reproduce:**

Follow the instructions here:
    [https://toil.readthedocs.io/en/latest/running/cloud/gce.html](https://toil.readthedocs.io/en/latest/running/cloud/gce.html)

Create a new service account with make bucket permissions and a new ssh key.  ^
These are the environment variables needed for running tests and credentialing (modify GOOGLE_APPLICATION_CREDENTIALS to your own path):
```
export GOOGLE_APPLICATION_CREDENTIALS=path to your downloaded service credentials json
export TOIL_TEST_INTEGRATIVE=True
export TRAVIS=True
export TOIL_GOOGLE_KEYNAME=name of your key
```
```
git clone https://github.com/DataBiosphere/toil.git
cd toil/src/toil/test/sort
python sort.py google:toil-dev:some-unique-name-like-jim
```

From what I discussed with @adamnovak and @DailyDreaming, I think there may be some type inconsistencies between Strings and Bytes that are being caused by Python2/Python3 compatibility problems.

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-763)
┆Issue Number: TOIL-763

### PR DESCRIPTION
**Pull request by jonathanxu18**
Should fix #3388.

### CODE REVIEW

File: src/toil/jobStores/googleJobStore.py [hunk 1]
```diff
@@ -452,8 +452,8 @@ def _writeFile(self, jobStoreID, fileObj, update=False, encrypt=True):
                 raise NoSuchFileException(jobStoreID)
         blob.upload_from_file(fileObj)
 
-    def _writeString(self, jobStoreID, stringToUpload, **kwarg):
-        self._writeFile(jobStoreID, StringIO(stringToUpload), **kwarg)
+    def _writeBytes(self, jobStoreID, stringToUpload, **kwarg):
```
Comment (DailyDreaming) [inline src/toil/jobStores/googleJobStore.py:455]: Could typing be added to this function?  Something like  `def _writeBytes(self, jobStoreID: str, stringToUpload: bytes, **kwarg): -> None`.

**Review by DailyDreaming — CHANGES_REQUESTED** [review]
Looks pretty good so far!  I just tried running this and it seems to have finished successfully, but it gave an error too?

```
Exception in thread Thread-11:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/quokka/git/toil/src/toil/statsAndLogging.py", line 176, in statsAndLoggingAggregator
    jobStore.writeStatsAndLogging(text)
  File "/home/quokka/git/toil/src/toil/jobStores/googleJobStore.py", line 367, in writeStatsAndLogging
    f.write(statsAndLoggingString)
TypeError: a bytes-like object is required, not 'str'

[2021-01-07T19:21:49-0800] [MainThread] [I] [toil.leader] Finished toil run successfully.

Workflow Progress 100%|█████████████████████████████████████████████████████████████| 31/31 (0 failures) [01:35<00:00, 0.33 jobs/s]
[2021-01-07T19:22:06-0800] [MainThread] [I] [toil.common] Successfully deleted the job store: <toil.jobStores.googleJobStore.GoogleJobStore object at 0x7f283a44f630>
```

I'm guessing that this might be stochastic?

This could probably also use a test.  Fortunately, I think that there already are some, we're just not running them.

I've added a couple of fixes based on running the tests.  I'll also add the tests since it involves setting up credentialing on the test infra (tests still running locally for me, but 27 passing so far; `toil.test.jobStores.jobStoreTest.GoogleJobStoreTest:testLargeFile` failed though).

```
export GOOGLE_APPLICATION_CREDENTIALS=/home/quokka/.credentials/google-toil-dev.json; export TOIL_TEST_INTEGRATIVE=True; export TRAVIS=True; export TOIL_GOOGLE_KEYNAME=toil2020; export TOIL_GOOGLE_PROJECTID=toil-dev

make test tests=src/toil/test/jobStores/jobStoreTest.py::GoogleJobStoreTest
```

Otherwise, this looks pretty good!  :slightly_smiling_face:

File: src/toil/jobStores/googleJobStore.py (hunk 1 above)
Comment (jonathanxu18) [reply src/toil/jobStores/googleJobStore.py:455]: Yep, I'll add that in!

**Comment by DailyDreaming** [comment]
@jonathanxu18 I've enabled tests.  If any fail, they can be skipped, and an issue made to fix this in a separate PR.  It looks like the majority are passing with this PR however.

**Comment by jonathanxu18** [comment]
> Looks pretty good so far! I just tried running this and it seems to have finished successfully, but it gave an error too?
> 
> ```
> Exception in thread Thread-11:
> Traceback (most recent call last):
>   File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
>     self.run()
>   File "/usr/lib/python3.6/threading.py", line 864, in run
>     self._target(*self._args, **self._kwargs)
>   File "/home/quokka/git/toil/src/toil/statsAndLogging.py", line 176, in statsAndLoggingAggregator
>     jobStore.writeStatsAndLogging(text)
>   File "/home/quokka/git/toil/src/toil/jobStores/googleJobStore.py", line 367, in writeStatsAndLogging
>     f.write(statsAndLoggingString)
> TypeError: a bytes-like object is required, not 'str'
> 
> [2021-01-07T19:21:49-0800] [MainThread] [I] [toil.leader] Finished toil run successfully.
> 
> Workflow Progress 100%|█████████████████████████████████████████████████████████████| 31/31 (0 failures) [01:35<00:00, 0.33 jobs/s]
> [2021-01-07T19:22:06-0800] [MainThread] [I] [toil.common] Successfully deleted the job store: <toil.jobStores.googleJobStore.GoogleJobStore object at 0x7f283a44f630>
> ```
> 
> I'm guessing that this might be stochastic?
> 
> This could probably also use a test. Fortunately, I think that there already are some, we're just not running them.
> 
> I've added a couple of fixes based on running the tests. I'll also add the tests since it involves setting up credentialing on the test infra (tests still running locally for me, but 27 passing so far; `toil.test.jobStores.jobStoreTest.GoogleJobStoreTest:testLargeFile` failed though).
> 
> ```
> export GOOGLE_APPLICATION_CREDENTIALS=/home/quokka/.credentials/google-toil-dev.json; export TOIL_TEST_INTEGRATIVE=True; export TRAVIS=True; export TOIL_GOOGLE_KEYNAME=toil2020; export TOIL_GOOGLE_PROJECTID=toil-dev
> 
> make test tests=src/toil/test/jobStores/jobStoreTest.py::GoogleJobStoreTest
> ```
> 
> Otherwise, this looks pretty good! 🙂

I was getting this error each time I was running it too but I was a bit confused since it was still passing successfully. I'll take a look and see if I can figure it out.

**Comment by DailyDreaming** [comment]
I rebased from master so that we could pull in the code to make use of the docker image caching so that it wouldn't error with too many pulls.

File: src/toil/jobStores/googleJobStore.py [hunk 2]
```diff
@@ -256,7 +256,7 @@ def writeFileStream(self, jobStoreID=None, cleanup=False, basename=None):
 
     def getEmptyFileStoreID(self, jobStoreID=None, cleanup=False, basename=None):
         fileID = self._newID(isFile=True, jobStoreID=jobStoreID if cleanup else None)
-        self._writeFile(fileID, StringIO(""))
+        self._writeFile(fileID, BytesIO(bytes("", 'utf-8')))
```
Comment (DailyDreaming) [inline src/toil/jobStores/googleJobStore.py:259]: I also want to add, you should only need `BytesIO(b"")` here.

Note: You're converting from unicode with this and it's equivalent to saying that the unicode string has encoding `utf-8`, please convert it to a raw bytestring based on that encoding:

`""` -> `bytes("", 'utf-8')` -> `b""`

**Comment by jonathanxu18** [comment]
> @jonathanxu18 I've enabled tests. If any fail, they can be skipped, and an issue made to fix this in a separate PR. It looks like the majority are passing with this PR however.

@DailyDreaming I made the changes that we talked about, but it looks like some of the Google jobstore tests are still failing. Do you think I should open a new issue to fix this in a separate PR?

File: src/toil/statsAndLogging.py [hunk 3]
```diff
@@ -173,7 +173,7 @@ def callback(fileHandle):
         # Finish the stats file
         text = json.dumps(dict(total_time=str(time.time() - startTime),
                                total_clock=str(get_total_cpu_time() - startClock)), ensure_ascii=True)
-        jobStore.writeStatsAndLogging(text)
+        jobStore.writeStatsAndLogging(bytes(text, 'utf-8'))
```
Comment (jonathanxu18) [inline src/toil/statsAndLogging.py:176]: @DailyDreaming I think this should fix the error we were getting at the end of workflows.

Comment (DailyDreaming) [reply src/toil/statsAndLogging.py:176]: That's great.  `writeStatsAndLogging` could probably also benefit from `type hinting` to describe that it takes `bytes` if you could add it!

**Review by DailyDreaming — APPROVED** [review]
Looks good!

**Pull request merged**